moonraker/components/wled.py

```diff
@@ -0,0 +1,502 @@
+            elif self.send_full_chain_data:
+                # Send a full set of color data (e.g. previous preset)
+                state = {"seg": {"i": []}}
+                cdata = []
+                for i in range(self.chain_count):
+                    idx = i * elem_size
+                    elem = []
+                    for p in self._chain_data[idx: idx+elem_size]:
+                        elem.append(p)
+                    cdata.append(elem)
+                state["seg"]["i"] = cdata
+                self.send_full_chain_data = False
+                await self._send_wled_command(state)
+            else:
+                # Only one pixel has changed so send just that one
+                elem = []
+                for p in self._chain_data[(index-1)*elem_size:
+                                          (index-1)*elem_size+elem_size]:
+                    elem.append(p)
+                await self._send_wled_command({"seg": {"i": [index, elem]}})
+        elif index is not None:
+            self.send_full_chain_data = True
+
+class WLED:
+    def __init__(self, config: ConfigHelper) -> None:
+        try:
+            # root_logger = logging.getLogger()
+            # root_logger.setLevel(logging.DEBUG)
+
+            self.server = config.get_server()
+            prefix_sections = config.get_prefix_sections("wled")
+            logging.info(f"WLED component loading strips: {prefix_sections}")
+            color_orders = {
+                "RGB": ColorOrder.RGB,
+                "RGBW": ColorOrder.RGBW
+            }
+            self.strips = {}
+            for section in prefix_sections:
+                cfg = config[section]
+
+                name_parts = cfg.get_name().split(maxsplit=1)
+                if len(name_parts) != 2:
+                    raise cfg.error(
+                        f"Invalid Section Name: {cfg.get_name()}")
+                name: str = name_parts[1]
+
+                logging.info(f"WLED strip: {name}")
+
+                color_order_cfg: str = cfg.get("color_order", "RGB")
+                color_order = color_orders.get(color_order_cfg)
+                if color_order is None:
+                    raise config.error(
+                        f"Color order not supported: {color_order_cfg}")
+
+                self.strips[name] = Strip(name, color_order, cfg)
+
+            event_loop = self.server.get_event_loop()
+            event_loop.register_callback(
+                self._initalize_strips, list(self.strips.values()))
+
+            # Register methods for GCODE remote methods
+            self.server.register_remote_method(
+                "wled_on", self.wled_on)
+            self.server.register_remote_method(
+                "wled", self.wled)
+            self.server.register_remote_method(
+                "wled_off", self.wled_off)
+            self.server.register_remote_method(
+                "set_wled", self.set_wled)
```

Comment:
As mentioned on Discord, I think that we need to reduce the number of remote methods registered.  The `set_led` method seems ok, it accepts parameters matching that of the SET_LED gcode.

I would recommend combining `wled_on` and `wled_off` into a single `set_wled_state()` that takes a `strip` and `state` argument.  It seems that a `preset` argument could be applied if present here.

The `wled()` method seems too low level to expose to Klipper via remote method.  Given that the wled supports presets, my suggestion would be to rely on it and expose the preset argument as mentioned above.

Reply:
OK sounds good, I'll back out the lower level controls, thanks for the feedback.

Reply:
Tried to make those changes now